// @unovis/vue@1.6.5/components/scatter/index.js
import { defineComponent as y, inject as m, computed as d, ref as F, onMounted as v, nextTick as b, onUnmounted as S, watch as u, openBlock as g, createElementBlock as f } from "vue";
import { Scatter as i } from "@unovis/ts";
import { useForwardProps as x, arePropsEqual as _ } from "../../utils/props.js";
import { componentAccessorKey as h } from "../../utils/context.js";
const k = { "data-vis-component": "" }, N = i.selectors, O = /* @__PURE__ */ y({
  __name: "index",
  props: {
    color: { type: [Function, String, Array, null] },
    size: { type: [Function, Number, null] },
    sizeScale: { type: [Object, Function] },
    sizeRange: {},
    shape: { type: [Function, String] },
    label: { type: [Function, String, null] },
    labelColor: { type: [Function, String, Array, null] },
    labelHideOverlapping: { type: Boolean },
    cursor: { type: [Function, String, null] },
    labelTextBrightnessRatio: {},
    labelPosition: { type: [Function, String, null] },
    strokeColor: { type: [Function, String, Array, null] },
    strokeWidth: { type: [Function, Number, null] },
    x: { type: [Function, Number, null] },
    y: { type: [Function, Number, null, Array] },
    id: { type: Function },
    xScale: { type: [Object, Function] },
    yScale: { type: [Object, Function] },
    excludeFromDomainCalculation: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(s, { expose: p }) {
    const n = m(h), r = s, o = d(() => n.data.value ?? r.data), a = x(r), t = F();
    return v(() => {
      b(() => {
        var e;
        t.value = new i(a.value), (e = t.value) == null || e.setData(o.value), n.update(t.value);
      });
    }), S(() => {
      var e;
      (e = t.value) == null || e.destroy(), n.destroy();
    }), u(a, (e, l) => {
      var c;
      _(e, l) || (c = t.value) == null || c.setConfig(a.value);
    }), u(o, () => {
      var e;
      (e = t.value) == null || e.setData(o.value);
    }), p({
      component: t
    }), (e, l) => (g(), f("div", k));
  }
});
export {
  N as VisScatterSelectors,
  O as default
};
