// @unovis/vue@1.6.5/components/sankey/index.js
import { defineComponent as d, inject as y, computed as b, ref as g, onMounted as m, nextTick as F, onUnmounted as S, watch as u, openBlock as h, createElementBlock as v } from "vue";
import { Sankey as c } from "@unovis/ts";
import { useForwardProps as x, arePropsEqual as T } from "../../utils/props.js";
import { componentAccessorKey as f } from "../../utils/context.js";
const k = { "data-vis-component": "" }, L = c.selectors, M = /* @__PURE__ */ d({
  __name: "index",
  props: {
    id: { type: Function },
    heightNormalizationCoeff: {},
    zoomScale: {},
    zoomPan: {},
    enableZoom: { type: Boolean },
    zoomExtent: {},
    zoomMode: {},
    exitTransitionType: {},
    enterTransitionType: {},
    highlightSubtreeOnHover: { type: Boolean },
    highlightDuration: {},
    highlightDelay: {},
    iterations: {},
    nodeSort: { type: [Function, null] },
    linkSort: { type: [Function, null] },
    nodeWidth: {},
    nodeAlign: {},
    nodeHorizontalSpacing: {},
    nodeMinHeight: {},
    nodeMaxHeight: {},
    nodePadding: {},
    showSingleNode: { type: Boolean },
    nodeCursor: { type: [Function, String, null] },
    nodeIcon: { type: [Function, String, null] },
    nodeColor: { type: [Function, String, Array, null] },
    nodeFixedValue: { type: [Function, Number, null] },
    nodeIconColor: { type: [Function, String, Array, null] },
    linkColor: { type: [Function, String, null] },
    linkValue: { type: [Function, Number, null] },
    linkCursor: { type: [Function, String, null] },
    label: { type: [Function, String, null] },
    subLabel: { type: [Function, String, null] },
    labelPosition: { type: [Function, String, null] },
    labelVerticalAlign: {},
    labelBackground: { type: Boolean },
    labelFit: {},
    labelMaxWidth: {},
    labelMaxWidthTakeAvailableSpace: { type: Boolean },
    labelMaxWidthTakeAvailableSpaceTolerance: {},
    labelExpandTrimmedOnHover: { type: Boolean },
    labelTrimMode: {},
    labelFontSize: {},
    labelTextSeparator: {},
    labelTextDecoration: { type: [Function, String, null] },
    labelForceWordBreak: { type: Boolean },
    labelColor: { type: [Function, String, Array, null] },
    labelCursor: { type: [Function, String, null] },
    labelVisibility: { type: Function },
    subLabelFontSize: {},
    subLabelColor: { type: [Function, String, Array, null] },
    subLabelPlacement: {},
    subLabelTextDecoration: { type: [Function, String, null] },
    subLabelToLabelInlineWidthRatio: {},
    onZoom: { type: Function },
    onLayoutCalculated: { type: Function },
    selectedNodeIds: {},
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(p, { expose: s }) {
    const o = y(f), a = p, t = b(() => o.data.value ?? a.data), l = x(a), n = g();
    return m(() => {
      F(() => {
        var e;
        n.value = new c(l.value), (e = n.value) == null || e.setData(t.value), o.update(n.value);
      });
    }), S(() => {
      var e;
      (e = n.value) == null || e.destroy(), o.destroy();
    }), u(l, (e, i) => {
      var r;
      T(e, i) || (r = n.value) == null || r.setConfig(l.value);
    }), u(t, () => {
      var e;
      (e = n.value) == null || e.setData(t.value);
    }), s({
      component: n
    }), (e, i) => (h(), v("div", k));
  }
});
export {
  L as VisSankeySelectors,
  M as default
};
