// @unovis/ts@1.6.5/components/topojson-map/utils.js
import { max, min } from 'd3-array';
import Supercluster from 'supercluster';
import { circlePath, polygon } from '../../utils/path.js';
import { getNumber, getString, clamp } from '../../utils/data.js';
import { getColor } from '../../utils/color.js';
import { estimateStringPixelLength } from '../../utils/text.js';
import { rectIntersect } from '../../utils/misc.js';
import { smartTransition } from '../../utils/d3.js';
import { TopoJSONMapPointShape } from './types.js';

function getLonLat(d, pointLongitude, pointLatitude) {
    const lat = getNumber(d, pointLatitude);
    const lon = getNumber(d, pointLongitude);
    return [lon, lat];
}
function arc(source, target, curvature) {
    if (!target || !source)
        return 'M0,0,l0,0z';
    const d = 3;
    const angleOffset = curvature || 0;
    const s = { x: source[0], y: source[1] };
    const t = { x: target[0], y: target[1] };
    const ds = { x: (t.x - s.x) / d, y: (t.y - s.y) / d };
    const dt = { x: (s.x - t.x) / d, y: (s.y - t.y) / d };
    let angle = 0.16667 * Math.PI * (1 + angleOffset);
    if (s.x < t.x)
        angle = -angle;
    const cs = Math.cos(angle);
    const ss = Math.sin(angle);
    const ct = Math.cos(-angle);
    const st = Math.sin(-angle);
    const dds = { x: (cs * ds.x) - (ss * ds.y), y: (ss * ds.x) + (cs * ds.y) };
    const ddt = { x: (ct * dt.x) - (st * dt.y), y: (st * dt.x) + (ct * dt.y) };
    return `M${s.x},${s.y} C${s.x + dds.x},${s.y + dds.y} ${t.x + ddt.x},${t.y + ddt.y} ${t.x},${t.y}`;
}
function getPointPathData({ x, y }, radius, shape) {
    switch (shape) {
        case TopoJSONMapPointShape.Triangle:
            return polygon(radius * 2, 3);
        case TopoJSONMapPointShape.Square:
            return polygon(radius * 2, 4);
        case TopoJSONMapPointShape.Circle:
        case TopoJSONMapPointShape.Ring:
        default:
            return circlePath(x, y, radius);
    }
}
function collideAreaLabels(selection, duration) {
    if (selection.size() === 0)
        return;
    const labelNodes = selection.nodes();
    const labelData = selection.data();
    // Reset all labels to visible
    labelNodes.forEach((node) => {
        node._labelVisible = true;
    });
    // Get actual font size
    let actualFontSize = 12;
    if (labelNodes.length > 0) {
        const computedStyle = getComputedStyle(labelNodes[0]);
        const fontSize = computedStyle.fontSize;
        if (fontSize)
            actualFontSize = parseFloat(fontSize);
    }
    const getBBox = (labelData) => {
        const [x, y] = labelData.centroid;
        const labelText = labelData.labelText || '';
        const width = estimateStringPixelLength(labelText, actualFontSize, 0.6);
        const height = actualFontSize;
        return {
            x: x - width / 2,
            y: y - height / 2,
            width,
            height,
        };
    };
    labelNodes.forEach((node1, i) => {
        const data1 = labelData[i];
        if (!node1._labelVisible)
            return;
        const label1BoundingRect = getBBox(data1);
        for (let j = 0; j < labelNodes.length; j++) {
            if (i === j)
                continue;
            const node2 = labelNodes[j];
            const data2 = labelData[j];
            if (!node2._labelVisible)
                continue;
            const label2BoundingRect = getBBox(data2);
            const intersect = rectIntersect(label1BoundingRect, label2BoundingRect, 0.25);
            if (intersect) {
                if (data1.area >= data2.area) {
                    node2._labelVisible = false;
                }
                else {
                    node1._labelVisible = false;
                    break;
                }
            }
        }
    });
    smartTransition(selection, duration)
        .style('opacity', (_d, i) => labelNodes[i]._labelVisible ? 1 : 0);
}
/** Get bounding rect of an SVG element in screen coordinates */
function getScreenBBox(el) {
    var _a, _b;
    const bbox = el.getBBox();
    const ctm = (_b = (_a = el).getScreenCTM) === null || _b === void 0 ? void 0 : _b.call(_a);
    if (!ctm || !el.ownerSVGElement) {
        return { x: bbox.x, y: bbox.y, width: bbox.width, height: bbox.height };
    }
    const pt = el.ownerSVGElement.createSVGPoint();
    const corners = [
        [bbox.x, bbox.y],
        [bbox.x + bbox.width, bbox.y],
        [bbox.x, bbox.y + bbox.height],
        [bbox.x + bbox.width, bbox.y + bbox.height],
    ];
    const transformed = corners.map(([x, y]) => {
        pt.x = x;
        pt.y = y;
        return pt.matrixTransform(ctm);
    });
    const xs = transformed.map(p => p.x);
    const ys = transformed.map(p => p.y);
    return {
        x: Math.min(...xs),
        y: Math.min(...ys),
        width: Math.max(...xs) - Math.min(...xs),
        height: Math.max(...ys) - Math.min(...ys),
    };
}
function collidePointBottomLabels(selection, duration) {
    if (selection.size() === 0)
        return;
    const labelNodes = selection.nodes();
    const labelData = selection.data();
    labelNodes.forEach((node, i) => {
        const d = labelData[i];
        node._labelVisible = !d.expandedClusterPoint;
    });
    labelNodes.forEach((node1, i) => {
        if (!node1._labelVisible)
            return;
        const data1 = labelData[i];
        let rect1;
        try {
            rect1 = getScreenBBox(node1);
        }
        catch (_a) {
            return;
        }
        for (let j = 0; j < labelNodes.length; j++) {
            if (i === j)
                continue;
            const node2 = labelNodes[j];
            if (!node2._labelVisible)
                continue;
            const data2 = labelData[j];
            let rect2;
            try {
                rect2 = getScreenBBox(node2);
            }
            catch (_b) {
                continue;
            }
            const intersect = rectIntersect(rect1, rect2, 2);
            if (intersect) {
                if (data1.radius >= data2.radius) {
                    node2._labelVisible = false;
                }
                else {
                    node1._labelVisible = false;
                    break;
                }
            }
        }
    });
    smartTransition(selection, duration)
        .attr('visibility', (_d, i) => labelNodes[i]._labelVisible ? null : 'hidden')
        .style('opacity', (_d, i) => labelNodes[i]._labelVisible ? 1 : 0);
}
function getDonutData(d, colorMap) {
    if (!colorMap || Object.keys(colorMap).length === 0) {
        return [];
    }
    return Object.keys(colorMap)
        .map(key => {
        const keyTyped = key;
        const config = colorMap[keyTyped];
        const value = Number(d[key]) || 0;
        return {
            name: key,
            value,
            color: (config === null || config === void 0 ? void 0 : config.color) || '#000',
            className: config === null || config === void 0 ? void 0 : config.className,
        };
    });
}
function toGeoJSONPoint(d, i, pointLatitude, pointLongitude) {
    const lat = getNumber(d, pointLatitude);
    const lon = getNumber(d, pointLongitude);
    return {
        type: 'Feature',
        properties: Object.assign(Object.assign({}, d), { _index: i }),
        geometry: {
            type: 'Point',
            coordinates: [lon, lat],
        },
    };
}
function calculateClusterIndex(data, config, maxClusterZoomLevel = 24) {
    const { colorMap, pointShape, latitude, longitude, clusteringDistance } = config;
    return new Supercluster({
        radius: clusteringDistance,
        maxZoom: maxClusterZoomLevel,
        minZoom: 0,
        map: (d) => {
            const shape = getString(d, pointShape);
            const clusterPoint = { shape };
            const dRecord = d;
            for (const key of Object.keys(colorMap || {})) {
                clusterPoint[key] = dRecord[key] || 0;
            }
            return clusterPoint;
        },
        reduce: (acc, clusterPoint) => {
            var _a, _b;
            acc.shape = acc.shape === clusterPoint.shape ? acc.shape : TopoJSONMapPointShape.Circle;
            acc.value = ((_a = acc.value) !== null && _a !== void 0 ? _a : 0) + ((_b = clusterPoint.value) !== null && _b !== void 0 ? _b : 0);
            for (const key of Object.keys(colorMap || {})) {
                acc[key] += clusterPoint[key];
            }
        },
    }).load(data.map((d, i) => toGeoJSONPoint(d, i, latitude, longitude)));
}
/* eslint-enable @typescript-eslint/naming-convention */
function getPointRadius(geoPoint, pointRadius, zoomLevel) {
    var _a, _b;
    const isDynamic = !pointRadius;
    const radius = isDynamic
        ? 1 + 2 * Math.pow(zoomLevel, 0.80)
        : getNumber(geoPoint.properties, pointRadius);
    const isCluster = geoPoint.properties.cluster;
    const clusterProps = (isCluster ? geoPoint.properties : null);
    const pointCount = (_b = (_a = clusterProps === null || clusterProps === void 0 ? void 0 : clusterProps.point_count) !== null && _a !== void 0 ? _a : geoPoint.properties.pointCount) !== null && _b !== void 0 ? _b : 1;
    return (isCluster && isDynamic)
        ? Math.max(Math.pow(pointCount, 0.35) * radius, radius * 1.1)
        : radius;
}
function geoJsonPointToScreenPoint(geoPoint, i, projection, config, zoomLevel) {
    var _a, _b;
    const isCluster = geoPoint.properties.cluster;
    const pos = projection(geoPoint.geometry.coordinates);
    const x = pos[0];
    const y = pos[1];
    const id = isCluster
        ? `cluster-${geoPoint.id}`
        : ((_b = (_a = (geoPoint.id !== undefined ? String(geoPoint.id) : undefined)) !== null && _a !== void 0 ? _a : getString(geoPoint.properties, config.pointId)) !== null && _b !== void 0 ? _b : geoPoint.geometry.coordinates.join(''));
    const pointColor = getColor(geoPoint.properties, (isCluster ? config.clusterColor : config.pointColor));
    const radius = getPointRadius(geoPoint, isCluster ? config.clusterRadius : config.pointRadius, zoomLevel);
    const shape = isCluster
        ? TopoJSONMapPointShape.Circle
        : (getString(geoPoint.properties, config.pointShape) || TopoJSONMapPointShape.Circle);
    const isRing = shape === TopoJSONMapPointShape.Ring;
    const donutData = getDonutData(geoPoint.properties, config.colorMap);
    const maxValue = max(donutData, d => d.value);
    const maxValueIndex = donutData.map(d => d.value).indexOf(maxValue || 0);
    const biggestDatum = donutData[maxValueIndex !== null && maxValueIndex !== void 0 ? maxValueIndex : 0];
    const color = isCluster
        ? pointColor
        : (isRing ? pointColor : (pointColor !== null && pointColor !== void 0 ? pointColor : biggestDatum === null || biggestDatum === void 0 ? void 0 : biggestDatum.color));
    const bbox = { x1: x - radius, y1: y - radius, x2: x + radius, y2: y + radius };
    const path = getPointPathData({ x: 0, y: 0 }, radius, shape);
    const _zIndex = 0;
    const screenPoint = Object.assign(Object.assign({}, geoPoint), { id,
        bbox,
        radius,
        donutData,
        path,
        color,
        isCluster, clusterIndex: isCluster ? geoPoint.properties.clusterIndex : undefined, _zIndex });
    return screenPoint;
}
/* eslint-enable @typescript-eslint/naming-convention */
function getClustersAndPoints(clusterIndex, bounds, zoom) {
    const points = clusterIndex.getClusters(bounds, zoom);
    for (const p of points) {
        const point = p;
        const isCluster = point.properties.cluster;
        if (isCluster) {
            const superclusterProps = point.properties;
            point.properties.clusterId = superclusterProps.cluster_id;
            point.properties.pointCount = superclusterProps.point_count;
            point.properties.pointCountAbbreviated = superclusterProps.point_count_abbreviated;
            point.properties.clusterIndex = clusterIndex;
            point.properties.clusterPoints = clusterIndex.getLeaves(superclusterProps.cluster_id, Infinity).map(d => d.properties);
        }
    }
    return points;
}
const getNextZoomLevelOnClusterClick = (level, maxZoom = 12) => clamp(1 + level * 1.5, level, maxZoom);
function shouldClusterExpand(cluster, zoomLevel, midLevel = 4, maxLevel = 8, maxClusterZoomLevel = 23) {
    if (!cluster || !cluster.isCluster)
        return false;
    const clusterProps = cluster.properties;
    if (!clusterProps.clusterId || !cluster.clusterIndex)
        return false;
    const clusterExpansionZoomLevel = cluster.clusterIndex.getClusterExpansionZoom(clusterProps.clusterId);
    return zoomLevel >= maxLevel ||
        (zoomLevel >= midLevel && (clusterProps.pointCount < 20 || clusterExpansionZoomLevel >= maxClusterZoomLevel));
}
function getClusterRadius(expandedCluster) {
    const { points } = expandedCluster;
    const minX = min(points.map(d => d.dx - d.packedRadius));
    const maxX = max(points.map(d => d.dx + d.packedRadius));
    const minY = min(points.map(d => d.dy - d.packedRadius));
    const maxY = max(points.map(d => d.dy + d.packedRadius));
    return Math.sqrt(Math.pow((maxX - minX), 2) + Math.pow((maxY - minY), 2)) * 0.5;
}

export { arc, calculateClusterIndex, collideAreaLabels, collidePointBottomLabels, geoJsonPointToScreenPoint, getClusterRadius, getClustersAndPoints, getDonutData, getLonLat, getNextZoomLevelOnClusterClick, getPointPathData, getPointRadius, shouldClusterExpand, toGeoJSONPoint };
