// @unovis/vue@1.6.5/components/area/index.js
import { defineComponent as y, inject as m, computed as d, ref as v, onMounted as F, nextTick as f, onUnmounted as x, watch as u, openBlock as b, createElementBlock as g } from "vue";
import { Area as i } from "@unovis/ts";
import { useForwardProps as h, arePropsEqual as A } from "../../utils/props.js";
import { componentAccessorKey as _ } from "../../utils/context.js";
const B = { "data-vis-component": "" }, N = i.selectors, j = /* @__PURE__ */ y({
  __name: "index",
  props: {
    color: { type: [Function, String, Array, null] },
    curveType: {},
    baseline: { type: [Function, Number, null] },
    opacity: { type: [Function, Number, null] },
    cursor: { type: [Function, String, null] },
    line: { type: Boolean },
    lineColor: { type: [Function, String, Array, null] },
    lineWidth: {},
    lineDashArray: { type: [Function, Array, null] },
    minHeight1Px: { type: Boolean },
    minHeight: {},
    stackMinHeight: { type: Boolean },
    x: { type: [Function, Number, null] },
    y: { type: [Function, Number, null, Array] },
    id: { type: Function },
    xScale: { type: [Object, Function] },
    yScale: { type: [Object, Function] },
    excludeFromDomainCalculation: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(s, { expose: p }) {
    const n = m(_), r = s, o = d(() => n.data.value ?? r.data), a = h(r), t = v();
    return F(() => {
      f(() => {
        var e;
        t.value = new i(a.value), (e = t.value) == null || e.setData(o.value), n.update(t.value);
      });
    }), x(() => {
      var e;
      (e = t.value) == null || e.destroy(), n.destroy();
    }), u(a, (e, c) => {
      var l;
      A(e, c) || (l = t.value) == null || l.setConfig(a.value);
    }), u(o, () => {
      var e;
      (e = t.value) == null || e.setData(o.value);
    }), p({
      component: t
    }), (e, c) => (b(), g("div", B));
  }
});
export {
  N as VisAreaSelectors,
  j as default
};
