// @unovis/vue@1.6.5/components/topojson-map/index.js
import { defineComponent as y, inject as m, computed as F, ref as d, onMounted as g, nextTick as S, onUnmounted as b, watch as a, openBlock as N, createElementBlock as v } from "vue";
import { TopoJSONMap as c } from "@unovis/ts";
import { useForwardProps as L, arePropsEqual as f } from "../../utils/props.js";
import { componentAccessorKey as C } from "../../utils/context.js";
const h = { "data-vis-component": "" }, M = c.selectors, w = /* @__PURE__ */ y({
  __name: "index",
  props: {
    projection: { type: [Function, Object] },
    topojson: {},
    mapFeatureName: {},
    mapFitToPoints: { type: Boolean },
    zoomFactor: {},
    zoomToLocation: {},
    disableZoom: { type: Boolean },
    zoomExtent: {},
    zoomDuration: {},
    linkWidth: { type: [Function, Number, null] },
    linkColor: { type: [Function, String, Array, null] },
    linkCursor: { type: [Function, String, null] },
    linkId: { type: [Function, String, null] },
    linkSource: { type: Function },
    linkTarget: { type: Function },
    sourceLongitude: { type: [Function, Number, null] },
    sourceLatitude: { type: [Function, Number, null] },
    targetLongitude: { type: [Function, Number, null] },
    targetLatitude: { type: [Function, Number, null] },
    sourcePointRadius: { type: [Function, Number, null] },
    sourcePointColor: { type: [Function, String, Array, null] },
    flowParticleColor: { type: [Function, String, Array, null] },
    flowParticleRadius: { type: [Function, Number, null] },
    flowParticleSpeed: { type: [Function, Number, null] },
    flowParticleDensity: { type: [Function, Number, null] },
    enableFlowAnimation: { type: Boolean },
    onSourcePointClick: { type: Function },
    onSourcePointMouseEnter: { type: Function },
    onSourcePointMouseLeave: { type: Function },
    areaId: { type: [Function, String, null] },
    areaColor: { type: [Function, String, Array, null] },
    areaCursor: { type: [Function, String, null] },
    areaLabel: { type: [Function, String, null] },
    pointColor: { type: [Function, String, Array, null] },
    pointRadius: { type: [Function, Number, null] },
    pointStrokeWidth: { type: [Function, Number, null] },
    pointShape: { type: [Function, String, null] },
    pointRingWidth: { type: [Function, Number, null] },
    pointCursor: { type: [Function, String, null] },
    longitude: { type: [Function, Number, null] },
    latitude: { type: [Function, Number, null] },
    pointLabel: { type: [Function, String, null] },
    pointLabelColor: { type: [Function, String, Array, null] },
    pointLabelPosition: {},
    pointBottomLabel: { type: [Function, String, null] },
    pointLabelTextBrightnessRatio: {},
    pointId: { type: Function },
    clusterColor: { type: [Function, String, Array, null] },
    clusterRadius: { type: [Function, Number, null] },
    clusterLabel: { type: [Function, String, null] },
    clusterLabelColor: { type: [Function, String, null] },
    clusterBottomLabel: { type: [Function, String, null] },
    clusterRingWidth: {},
    clusterBackground: { type: Boolean },
    clusterExpandOnClick: { type: Boolean },
    clusteringDistance: {},
    clustering: { type: Boolean },
    heatmapMode: { type: Boolean },
    heatmapModeBlurStdDeviation: {},
    heatmapModeZoomLevelThreshold: {},
    colorMap: {},
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(p, { expose: s }) {
    const o = m(C), l = p, e = F(() => o.data.value ?? l.data), u = L(l), n = d();
    return g(() => {
      S(() => {
        var t;
        n.value = new c(u.value), (t = n.value) == null || t.setData(e.value), o.update(n.value);
      });
    }), b(() => {
      var t;
      (t = n.value) == null || t.destroy(), o.destroy();
    }), a(u, (t, i) => {
      var r;
      f(t, i) || (r = n.value) == null || r.setConfig(u.value);
    }), a(e, () => {
      var t;
      (t = n.value) == null || t.setData(e.value);
    }), s({
      component: n
    }), (t, i) => (N(), v("div", h));
  }
});
export {
  M as VisTopoJSONMapSelectors,
  w as default
};
