// @unovis/vue@1.6.5/components/annotations/index.js
import { defineComponent as p, inject as u, ref as d, onMounted as m, nextTick as l, onUnmounted as f, watch as v, openBlock as _, createElementBlock as x } from "vue";
import { Annotations as a } from "@unovis/ts";
import { useForwardProps as B, arePropsEqual as y } from "../../utils/props.js";
import { annotationsAccessorKey as g } from "../../utils/context.js";
const h = { "data-vis-annotations": "" }, P = a.selectors, T = /* @__PURE__ */ p({
  __name: "index",
  props: {
    items: {},
    renderTextBoundingBoxes: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(c, { expose: i }) {
    const n = u(g), t = B(c), o = d();
    return m(() => {
      l(() => {
        o.value = new a(t.value), n.update(o.value);
      });
    }), f(() => {
      var e;
      (e = o.value) == null || e.destroy(), n.destroy();
    }), v(t, (e, s) => {
      var r;
      y(e, s) || (r = o.value) == null || r.setConfig(t.value);
    }), i({
      component: o
    }), (e, s) => (_(), x("div", h));
  }
});
export {
  P as VisAnnotationsSelectors,
  T as default
};
