// @unovis/vue@1.6.5/html-components/flow-legend/index.js
import { defineComponent as m, computed as u, ref as s, onMounted as p, nextTick as f, onUnmounted as v, watch as w, openBlock as g, createElementBlock as _ } from "vue";
import { FlowLegend as d } from "@unovis/ts";
import { useForwardProps as y, arePropsEqual as C } from "../../utils/props.js";
const b = d.selectors, P = /* @__PURE__ */ m({
  __name: "index",
  props: {
    margin: {},
    customWidth: {},
    items: {},
    spacing: {},
    lineColor: {},
    labelColor: {},
    labelFontSize: {},
    arrowSymbol: {},
    arrowColor: {},
    arrowSymbolYOffset: {},
    onLegendItemClick: { type: Function },
    renderIntoProvidedDomNode: { type: Boolean },
    data: {}
  },
  setup(c, { expose: i }) {
    const r = c;
    u(() => r.data);
    const n = y(r), o = s(), t = s();
    return p(() => {
      f(() => {
        t.value && (o.value = new d(t.value, { ...n.value, renderIntoProvidedDomNode: !0 }));
      });
    }), v(() => {
      var e;
      (e = o.value) == null || e.destroy();
    }), w(n, (e, l) => {
      var a;
      C(e, l) || (a = o.value) == null || a.setConfig(n.value);
    }), i({
      component: o
    }), (e, l) => (g(), _("div", {
      "data-vis-flow-legend": "",
      ref_key: "elRef",
      ref: t
    }, null, 512));
  }
});
export {
  b as VisFlowLegendSelectors,
  P as default
};
