// @unovis/vue@1.6.5/components/treemap/index.js
import { defineComponent as d, inject as m, computed as f, ref as y, onMounted as F, nextTick as b, onUnmounted as v, watch as s, openBlock as S, createElementBlock as B } from "vue";
import { Treemap as c } from "@unovis/ts";
import { useForwardProps as T, arePropsEqual as g } from "../../utils/props.js";
import { componentAccessorKey as L } from "../../utils/context.js";
const _ = { "data-vis-component": "" }, k = c.selectors, A = /* @__PURE__ */ d({
  __name: "index",
  props: {
    id: { type: Function },
    value: { type: [Function, Number, null] },
    layers: {},
    numberFormat: { type: Function },
    tileLabel: { type: Function },
    tileColor: { type: [Function, String, Array, null] },
    tileFunction: { type: Function },
    tileSort: { type: [Function, null] },
    tilePadding: {},
    tilePaddingTop: {},
    tileShowHtmlTooltip: { type: Boolean },
    labelInternalNodes: { type: Boolean },
    labelOffsetX: {},
    labelOffsetY: {},
    labelFit: {},
    labelTrimMode: {},
    tileBorderRadius: {},
    tileBorderRadiusFactor: {},
    enableLightnessVariance: { type: Boolean },
    enableTileLabelFontSizeVariation: { type: Boolean },
    tileLabelSmallFontSize: {},
    tileLabelMediumFontSize: {},
    tileLabelLargeFontSize: {},
    showTileClickAffordance: { type: Boolean },
    lightnessVariationAmount: {},
    minTileSizeForLabel: {},
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(u, { expose: p }) {
    const o = m(L), l = u, n = f(() => o.data.value ?? l.data), a = T(l), t = y();
    return F(() => {
      b(() => {
        var e;
        t.value = new c(a.value), (e = t.value) == null || e.setData(n.value), o.update(t.value);
      });
    }), v(() => {
      var e;
      (e = t.value) == null || e.destroy(), o.destroy();
    }), s(a, (e, i) => {
      var r;
      g(e, i) || (r = t.value) == null || r.setConfig(a.value);
    }), s(n, () => {
      var e;
      (e = t.value) == null || e.setData(n.value);
    }), p({
      component: t
    }), (e, i) => (S(), B("div", _));
  }
});
export {
  k as VisTreemapSelectors,
  A as default
};
