// @unovis/vue@1.6.5/html-components/leaflet-flow-map/index.js
import { defineComponent as d, computed as F, ref as a, onMounted as m, nextTick as f, onUnmounted as g, watch as c, openBlock as S, createElementBlock as b } from "vue";
import { LeafletFlowMap as p } from "@unovis/ts";
import { useForwardProps as v, arePropsEqual as L } from "../../utils/props.js";
const P = p.selectors, C = /* @__PURE__ */ d({
  __name: "index",
  props: {
    sourceLongitude: { type: [Function, Number, null] },
    sourceLatitude: { type: [Function, Number, null] },
    targetLongitude: { type: [Function, Number, null] },
    targetLatitude: { type: [Function, Number, null] },
    sourcePointRadius: { type: [Function, Number, null] },
    sourcePointColor: { type: [Function, String, Array, null] },
    flowParticleColor: { type: [Function, String, Array, null] },
    flowParticleRadius: { type: [Function, Number, null] },
    flowParticleSpeed: { type: [Function, Number, null] },
    flowParticleDensity: { type: [Function, Number, null] },
    onSourcePointClick: { type: Function },
    onSourcePointMouseEnter: { type: Function },
    onSourcePointMouseLeave: { type: Function },
    width: {},
    height: {},
    flyToDuration: {},
    fitViewPadding: {},
    zoomDuration: {},
    initialBounds: {},
    fitBoundsOnUpdate: {},
    fitViewOnInit: { type: Boolean },
    fitViewOnUpdate: { type: Boolean },
    style: {},
    styleDarkTheme: {},
    accessToken: {},
    attribution: {},
    renderer: {},
    onMapInitialized: { type: Function },
    onMapMoveZoom: { type: Function },
    onMapMoveStart: { type: Function },
    onMapMoveEnd: { type: Function },
    onMapZoomStart: { type: Function },
    onMapZoomEnd: { type: Function },
    onMapClick: { type: Function },
    pointLongitude: { type: [Function, Number, null] },
    pointLatitude: { type: [Function, Number, null] },
    pointId: { type: [Function, String, null] },
    pointShape: { type: [Function, String, null] },
    pointColor: { type: [Function, String, Array, null] },
    pointRadius: { type: [Function, Number, null] },
    pointLabel: { type: [Function, String, null] },
    pointLabelColor: { type: [Function, String, null] },
    pointBottomLabel: { type: [Function, String, null] },
    pointCursor: { type: [Function, String, null] },
    pointRingWidth: {},
    selectedPointId: {},
    clusterColor: { type: [Function, String, Array, null] },
    clusterRadius: { type: [Function, Number, null] },
    clusterLabel: { type: [Function, String, null] },
    clusterLabelColor: { type: [Function, String, null] },
    clusterBottomLabel: { type: [Function, String, null] },
    clusterRingWidth: {},
    clusterBackground: { type: Boolean },
    clusterExpandOnClick: { type: Boolean },
    clusteringDistance: {},
    colorMap: {},
    topoJSONLayer: {},
    tooltip: {},
    ariaLabel: {},
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(s, { expose: y }) {
    const i = s, e = F(() => i.data), o = v(i), n = a(), u = a();
    return m(() => {
      f(() => {
        u.value && (n.value = new p(u.value, o.value, e.value));
      });
    }), g(() => {
      var t;
      (t = n.value) == null || t.destroy();
    }), c(o, (t, l) => {
      var r;
      L(t, l) || (r = n.value) == null || r.setConfig(o.value);
    }), c(e, () => {
      var t;
      (t = n.value) == null || t.setData(e.value);
    }), y({
      component: n
    }), (t, l) => (S(), b("div", {
      "data-vis-leaflet-flow-map": "",
      ref_key: "elRef",
      ref: u
    }, null, 512));
  }
});
export {
  P as VisLeafletFlowMapSelectors,
  C as default
};
