// @unovis/ts@1.6.5/components/topojson-map/modules/flow.js
import { getNumber, isNumber, getString } from '../../../utils/data.js';
import { getColor } from '../../../utils/color.js';
import { arc } from '../utils.js';
import { flowParticle } from '../style.js';

// Utils
function initFlowFeatures(ctx) {
    var _a;
    const { config, datamodel } = ctx;
    // Use raw links data instead of processed links to avoid point lookup issues for flows
    const rawLinks = ((_a = datamodel.data) === null || _a === void 0 ? void 0 : _a.links) || [];
    // Clear existing flow data
    ctx._flowParticles = [];
    ctx._sourcePoints = [];
    if (!rawLinks || rawLinks.length === 0)
        return;
    // Create source points and flow particles for each link
    rawLinks.forEach((link, i) => {
        var _a, _b;
        // Try to get coordinates from flow-specific accessors first, then fall back to link endpoints
        let sourceLon;
        let sourceLat;
        let targetLon;
        let targetLat;
        if (config.sourceLongitude && config.sourceLatitude) {
            sourceLon = getNumber(link, config.sourceLongitude);
            sourceLat = getNumber(link, config.sourceLatitude);
        }
        else {
            // Fall back to using linkSource point coordinates
            const sourcePoint = (_a = config.linkSource) === null || _a === void 0 ? void 0 : _a.call(config, link);
            if (typeof sourcePoint === 'object' && sourcePoint !== null) {
                sourceLon = getNumber(sourcePoint, config.longitude);
                sourceLat = getNumber(sourcePoint, config.latitude);
            }
            else {
                return; // Skip if can't resolve source coordinates
            }
        }
        if (config.targetLongitude && config.targetLatitude) {
            targetLon = getNumber(link, config.targetLongitude);
            targetLat = getNumber(link, config.targetLatitude);
        }
        else {
            // Fall back to using linkTarget point coordinates
            const targetPoint = (_b = config.linkTarget) === null || _b === void 0 ? void 0 : _b.call(config, link);
            if (typeof targetPoint === 'object' && targetPoint !== null) {
                targetLon = getNumber(targetPoint, config.longitude);
                targetLat = getNumber(targetPoint, config.latitude);
            }
            else {
                return; // Skip if can't resolve target coordinates
            }
        }
        if (!isNumber(sourceLon) || !isNumber(sourceLat) || !isNumber(targetLon) || !isNumber(targetLat)) {
            return;
        }
        // Create source point
        const sourcePos = ctx._projection([sourceLon, sourceLat]);
        if (sourcePos) {
            const sourcePoint = {
                lat: sourceLat,
                lon: sourceLon,
                x: sourcePos[0],
                y: sourcePos[1],
                radius: getNumber(link, config.sourcePointRadius),
                color: getColor(link, config.sourcePointColor, i),
                flowData: link,
            };
            ctx._sourcePoints.push(sourcePoint);
        }
        // Use the same arc as _renderLinks for flow animation
        const sourceProj = ctx._projection([sourceLon, sourceLat]);
        const targetProj = ctx._projection([targetLon, targetLat]);
        if (!sourceProj || !targetProj)
            return;
        // Generate SVG arc path string using the same arc() function
        const arcPath = arc(sourceProj, targetProj);
        // Create a temporary SVG path element for sampling
        const tempPath = document.createElementNS('http://www.w3.org/2000/svg', 'path');
        tempPath.setAttribute('d', arcPath);
        const pathLength = tempPath.getTotalLength();
        const dist = Math.sqrt(Math.pow((targetLat - sourceLat), 2) + Math.pow((targetLon - sourceLon), 2));
        const numParticles = Math.max(1, Math.round(dist * getNumber(link, config.flowParticleDensity)));
        const velocity = getNumber(link, config.flowParticleSpeed);
        const radius = getNumber(link, config.flowParticleRadius);
        const color = getColor(link, config.flowParticleColor, i);
        for (let j = 0; j < numParticles; j += 1) {
            const progress = j / numParticles;
            const pt = tempPath.getPointAtLength(progress * pathLength);
            const particle = {
                x: pt.x,
                y: pt.y,
                velocity,
                radius,
                color,
                progress,
                arcPath,
                pathLength,
                id: `${getString(link, config.linkId, i) || i}-${j}`,
                flowData: undefined,
            };
            ctx._flowParticles.push(particle);
        }
    });
}
function updateFlowParticles(ctx) {
    if (ctx._flowParticles.length === 0)
        return;
    const zoomLevel = ctx._currentZoomLevel || 1;
    ctx._flowParticles.forEach(particle => {
        // Move particle along the arc path using progress
        particle.progress += particle.velocity * 0.01;
        if (particle.progress > 1)
            particle.progress = 0;
        // Use the stored SVG path and pathLength
        if (particle.arcPath && typeof particle.pathLength === 'number') {
            const tempPath = document.createElementNS('http://www.w3.org/2000/svg', 'path');
            tempPath.setAttribute('d', particle.arcPath);
            const pt = tempPath.getPointAtLength(particle.progress * particle.pathLength);
            particle.x = pt.x;
            particle.y = pt.y;
        }
    });
    // Update DOM elements directly without data rebinding (for performance)
    ctx._flowParticlesGroup
        .selectAll(`.${flowParticle}`)
        .attr('cx', (d, i) => { var _a; return ((_a = ctx._flowParticles[i]) === null || _a === void 0 ? void 0 : _a.x) || 0; })
        .attr('cy', (d, i) => { var _a; return ((_a = ctx._flowParticles[i]) === null || _a === void 0 ? void 0 : _a.y) || 0; })
        .attr('r', (d, i) => { var _a; return (((_a = ctx._flowParticles[i]) === null || _a === void 0 ? void 0 : _a.radius) || 1) / zoomLevel; });
}

export { initFlowFeatures, updateFlowParticles };
